File: core/java/src/main/java/hara/lang/protocol/IStream.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

/** Asynchronous pull source. A fulfilled null value denotes end-of-stream. */
@HaraProtocolBinding(
    namespace = "std.protocol.istream", name = "IStream", parents = {"IClose"})
public interface IStream extends IClose {
  @HaraMethod(value = "next", arity = 1)
  Object next();
}

File: core/rust/src/lang/protocol/istream.rs
use super::IClose;
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.istream",
    name = "IStream",
    parents = ["IClose"]
)]
pub trait IStream: IClose {
    type Item;

    #[hara_method(value = "next", arity = 1)]
    fn next(&mut self) -> Option<Self::Item>;
}
